apiserver/authentication/jwt/jwt.go

```diff
@@ -79,7 +80,26 @@ func NewAuthenticatorWithHTTPClient(
 
 // Authenticate implements EntityAuthenticator
 func (j *JWTAuthenticator) Authenticate(req *http.Request) (authentication.AuthInfo, error) {
-	return authentication.AuthInfo{}, fmt.Errorf("jwt authenticate %w", errors.NotImplemented)
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		return authentication.AuthInfo{}, errors.NotFoundf("authorization header missing")
+	}
```

Comment:
This should be a bad request. We might have to do some additional plumbing to get this to work.

Reply:
Yes, as described in the description, this change is to comply with the for loop `apiserver/stateauthenticator/auth.go:21`. If there is any error besides not-found or not-implemented the authentication fails and it doesn't try the next authenticator.
I can change and add the not-valid to the for logic, what do you think?

Reply:
Well, that's not great! 

I don't believe we can do that, as that breaks the compatibility guarantee of the API. All errors by definition are included in the API definition. As JAAS is the only consumer of this, maybe we could be flexible here, knowing that we could return BadRequest. Otherwise, we'll have to return NotFound and live with it.

Hence why we should really consider how the authentication is implemented and be sure we're doing the right things.

Reply:
Sorry, I shot the original comment too early.

I'm not sure where your link leads to?

https://github.com/juju/juju/blob/bc3a4ce49a9451a1d16ce07fb18332c14060b302/apiserver/stateauthenticator/auth.go#L21

Do you mean?

https://github.com/juju/juju/blob/bc3a4ce49a9451a1d16ce07fb18332c14060b302/apiserver/authentication/http.go#L20-L31

If we return `errors.BadRequest` then we need to understand what to do there. I believe if we get a `errors.BadRequest` then we should return that directly to the client.

Reply:
I just want to clarify, that we should ensure that we don't leak information about the user, just that JWT is malformed if the JWT header _was_ sent. Otherwise returning NotFound is valid.

Reply:
yes, it is the code I was referring to.

In this current implementation we return:
- not found if the Authorization header is missing -> go to next authentitcato
- not found if the Authorization header is there, but it is not formatted as "Bearer <>"  -> go to next authentitcato
- another error if the jwt cannot be parsed -> return the error to the user

Reply:
Ok, that's fine. I'm happy with that logic. Leave as not found error, given the scenarios you've described. Can you document the expectations of that in the `HTTPStrategicAuthenticator.Authenticate` method, and validate the tests work as expected according to those expectations?

Thanks for this @SimoneDutto

Reply:
perfect.

the behavior is described at the interface level, a few lines above the authenticate method
https://github.com/juju/juju/blob/bc3a4ce49a9451a1d16ce07fb18332c14060b302/apiserver/authentication/http.go#L13-L16

I've added some tests for the HTTPStrategicAuthenticator